atlasdb-backup/src/main/java/com/palantir/atlasdb/backup/AtlasRestoreService.java

```diff
@@ -173,79 +182,89 @@ public Set<Namespace> visitUnknown(String unknownType) {
+        Set<AtlasService> atlasServicesToRestore = getAtlasServicesToRestore(completedBackups);
 
         if (completedBackups.isEmpty()) {
             log.info(
                     "Attempted to complete restore, but no completed backups were found",
                     SafeArg.of("restoreRequests", restoreRequests));
             return ImmutableSet.of();
         } else if (completedBackups.size() < restoreRequests.size()) {
-            Set<Namespace> namespacesWithBackup = completedBackups.values().stream()
-                    .map(CompletedBackup::getNamespace)
+            Set<AtlasService> atlasServicesWithBackup = completedBackups.keySet().stream()
+                    .map(RestoreRequest::oldAtlasService)
                     .collect(Collectors.toSet());
-            Set<Namespace> allOldNamespaces =
-                    restoreRequests.stream().map(RestoreRequest::oldNamespace).collect(Collectors.toSet());
-            Set<Namespace> namespacesWithoutBackup = Sets.difference(allOldNamespaces, namespacesWithBackup);
+            Set<AtlasService> allOldAtlasServices = restoreRequests.stream()
+                    .map(RestoreRequest::oldAtlasService)
+                    .collect(Collectors.toSet());
+            Set<AtlasService> atlasServicesWithoutBackup =
+                    Sets.difference(allOldAtlasServices, atlasServicesWithBackup);
             log.warn(
-                    "Completed backups were not found for some namespaces",
-                    SafeArg.of("namespacesWithBackup", namespacesWithBackup),
-                    SafeArg.of("namespacesWithoutBackup", namespacesWithoutBackup));
+                    "Completed backups were not found for some atlasServices",
+                    SafeArg.of("atlasServicesWithBackup", atlasServicesWithBackup),
+                    SafeArg.of("atlasServicesWithoutBackup", atlasServicesWithoutBackup));
         }
 
         // Fast forward timestamps
         Map<Namespace, CompletedBackup> completeRequest = KeyedStream.stream(completedBackups)
-                .mapKeys(RestoreRequest::newNamespace)
+                .mapKeys(RestoreRequest::newAtlasService)
+                .mapKeys(AtlasService::getNamespace)
+                .collectToMap();
+        Map<Namespace, AtlasService> namespaceToServices = KeyedStream.of(atlasServicesToRestore)
+                .mapKeys(AtlasService::getNamespace)
                 .collectToMap();
+
         CompleteRestoreResponse response =
                 atlasRestoreClient.completeRestore(authHeader, CompleteRestoreRequest.of(completeRequest));
-        Set<Namespace> successfulNamespaces = response.getSuccessfulNamespaces();
-        Set<Namespace> failedNamespaces = Sets.difference(namespacesToRestore, successfulNamespaces);
-        if (!failedNamespaces.isEmpty()) {
+        Set<AtlasService> successfulAtlasServices = response.getSuccessfulNamespaces().stream()
+                .map(namespaceToServices::get)
```

Comment:
is this guaranteed to be present in the map?

Reply:
Yes - the response to `completeRestore` should be a subset of the requested namespaces.